src/mesh/parallel_tree.jl

```diff
@@ -0,0 +1,833 @@
+
+# Composite type that represents a NDIMS-dimensional tree.
+#
+# Implements everything required for AbstractContainer.
+#
+# Note: The way the data structures are set up and the way most algorithms
+# work, it is *always* assumed that
+#   a) we have a balanced tree (= at most one level difference between
+#                                 neighboring cells, or 2:1 rule)
+#   b) we may not have all children (= some children may not exist)
+#   c) the tree is stored depth-first
+#
+# However, the way the refinement/coarsening algorithms are currently
+# implemented, we only have fully refined cells. That is, a cell either has 2^NDIMS children or
+# no children at all (= leaf cell). This restriction is also assumed at
+# multiple positions in the refinement/coarsening algorithms.
+#
+# An exception to the 2:1 rule exists for the low-level `refine_unbalanced!`
+# function, which is required for implementing level-wise refinement in a sane
+# way. Also, depth-first ordering *might* not by guaranteed during
+# refinement/coarsening operations.
+mutable struct ParallelTree{NDIMS} <: AbstractContainer
+  parent_ids::Vector{Int}
+  child_ids::Matrix{Int}
+  neighbor_ids::Matrix{Int}
+  levels::Vector{Int}
+  coordinates::Matrix{Float64}
+  original_cell_ids::Vector{Int}
+  domain_ids::Vector{Int}
+
+  capacity::Int
+  length::Int
+  dummy::Int
+
+  center_level_0::MVector{NDIMS, Float64}
```

Comment:
Do you need to modify elements of `center_level_0`? Otherwise, we could also use an `SVector`.

Reply:
Fixed in 9a3b277